### ISSUE TITLE
SaveAs function only fills File.Path value later, causing problems with subsequent use of Save function

### ISSUE THREAD

**Issue by Orionsg** [360EntSecGroup-Skylar/excelize#872]
Apparently, the SaveAs function's filling of the File.Path field is done concurrently rather than right away before the execution of code after the SaveAs function is called. This can cause problems when the subsequent code uses the Save function, giving the error:
"Could not save the Excel file `test.xlsx`: no path defined for file, consider File.WriteTo or File.Write
exit status 1"

The test program below contains a "saveXlFile" function which is called several times after the "SaveAs" function is initially used to save a new Excel file "test.xlsx". Line 30 and 37 containing calls to saveXlFile have been remarked out as they otherwise cause the error mentioned above  However when running the program with the lines 30 and 37 included in the execution (remove the comment characters //), it varies which of them causes the error, and hence my assumption that the SaveAs processing is being executed in parallel.

Here is the test program:

```
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/360EntSecGroup-Skylar/excelize/v2"
)

const s_file = "test.xlsx"
const s_sh_test = "test"

func main() {
	var wb *excelize.File
	var err error

	if fileExists(s_file) {
		wb, err = excelize.OpenFile(s_file)
		if err != nil {
			log.Fatal(err)
		}
	} else {
		wb = excelize.NewFile()
		err = wb.SaveAs(s_file)
		if err != nil {
			log.Fatal("Could not save new Excel file `", s_file, "`: ", err)
		}
		// fmt.Println("Immediate save of new Excel file after SaveAs")
		// saveXlFile(wb)
	}
	if sheetExists(wb, s_sh_test) {
		wb.DeleteSheet(s_sh_test)
	}
	wb.NewSheet(s_sh_test)
	// fmt.Println("Saving Excel file with new sheet test")
	// saveXlFile(wb)

	// add values to the test sheet
	sl_cells := []string{"A1", "A2", "A3", "B1", "B2", "B3", "C1", "C2", "C3"}
	for _, s_cell := range sl_cells {
		wb.SetCellStr(s_sh_test, s_cell, s_cell)
	}
	fmt.Println("Saving Excel file with cells filled in test sheet")
	saveXlFile(wb)

	s_ch := ""
	sl_menu := []string{"yes", "no"}

	for !contains(sl_menu, s_ch) {
		fmt.Printf("Read content of sheet `%s` using GetRows? [yes/no] \n", s_sh_test)
		fmt.Scanln(&s_ch)
	}
	if s_ch == "yes" {
		fmt.Printf("Content of sheet `%s`: \n", s_sh_test)
		rows, err := wb.GetRows(s_sh_test)
		if err != nil {
			log.Fatal(err)
			return
		}
		for i_row, row := range rows {
			fmt.Print("Row: ", i_row, "\t")
			for _, s_cell := range row {
				if s_cell == "" {
					s_cell = "[]"
				}
				fmt.Print(s_cell, "\t")
			}
			fmt.Println()
		}
	}
	saveXlFile(wb)
	fmt.Println("Final save of Excel file")
}

func saveXlFile(wb *excelize.File) {
	err := wb.Save()
	if err != nil {
		log.Fatal("Could not save the Excel file `", s_file, "`: ", err)
	}
}

func sheetExists(wb *excelize.File, s_sh_name string) bool {
	b_fnd := false
	for _, s_sh := range wb.GetSheetList() {
		if s_sh == s_sh_name {
			b_fnd = true
			break
		}
	}
	return b_fnd
}

func contains(sl_str []string, ss string) bool {
	if len(sl_str) == 0 || ss == "" {
		return false
	}
	for _, str := range sl_str {
		if str == ss {
			return true
		}
	}
	return false
}

func fileExists(s_file string) bool {
	info, err := os.Stat(s_file)
	if os.IsNotExist(err) {
		return false
	}
	if info.IsDir() {
		log.Fatal("Error: ", s_file, " is a directory")
	}
	return true
}
```

**Comment by xuri** [comment]
If you create a new spreadsheet by `NewFile`, it only can be saving it by `SaveAs`. If you open an existing spreadsheet by `OpenFile`, you can save the change by `Save` or saved it as another file by `SaveAs`.

**Comment by Orionsg** [comment]
Look again at my code, please.
I check if the Excel file exists, in which case I use the OpenFile. However, if it does not exist, I use NewFile to create it and immediately afterwards use SaveAs to save it with the given name. At that point the file does then exists. Later in the program, I should not now able to use the Save function as it has now been created by the SaveAs. However, there is a problem with that. which is what my bug report is about.

**Comment by Orionsg** [comment]
Here is shorter program that focus on just this issue, which gives this error message when using the master branch:
"Could not save the Excel file `test.xlsx`: no path defined for file, consider File.WriteTo or File.Write".

After already using SaveAs,  one should now be able to use Save, correct?
Otherwise, how can one do further changes on the Excel file and then save those?

```
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/360EntSecGroup-Skylar/excelize/v2"
)

const s_file = "test.xlsx"

func main() {
	var wb *excelize.File
	var err error

	if fileExists(s_file) {
		err = os.Remove(s_file)
		if err != nil {
			log.Fatal("Problem removing file `", s_file, "`: ", err)
		}
	}
	wb = excelize.NewFile()
	err = wb.SaveAs(s_file)
	if err != nil {
		log.Fatal("Could not save new Excel file `", s_file, "`: ", err)
	}

	// now I should be able to save this new file as I have already used SaveAs
	fmt.Println("Trying to Save after use of SaveAs")
	saveXlFile(wb)
}

func saveXlFile(wb *excelize.File) {
	err := wb.Save()
	if err != nil {
		log.Fatal("Could not save the Excel file `", s_file, "`: ", err)
	}
}

func fileExists(s_file string) bool {
	info, err := os.Stat(s_file)
	if os.IsNotExist(err) {
		return false
	}
	if info.IsDir() {
		log.Fatal("Error: ", s_file, " is a directory")
	}
	return true
}
```

**Comment by Orionsg** [comment]
Perhaps, as you say, when creating a new Excel with NewFile, one can only use SaveAs, even after the file has been saved using that same function. However, this seems very counter-intuitive and different from other Excel packages and programs I have seen that handle documents. 

I would say that only being able to use SaveAs, even when already having saved the new file (it is then not really a new file any longer), is so against convention that many new users of the Excelize will very likely run into this issue again and again. Furthermore, it adds unnecessary logic to an Excelize program, which can handle both new and existing files, having to keep track of whether a file started as new or was an existing file being opened, even if you add your own  "save" function that keeps track of the state of the Excel file.

Is it not possible that when SaveAs is used, the Excel.File struct is updated so it is identical to the one created by the OpenFile function?

**Comment by xuri** [comment]
To resolve this issue, we just set saved path in the [`SaveAs`](https://github.com/xuri/excelize/blob/544ef18a8cb9949fcb8833c6d2816783c90f3318/file.go#L67) like this:

```go
func (f *File) SaveAs(name string, opt ...Options) error {
    f.Path = name
    if len(name) > MaxFileNameLength {
        return ErrMaxFileNameLength
    }
    ...
}
```

Would you like to make PR for this? or I'll fix it later.

**Comment by xuri** [comment]
I have fixed this issue, please try to upgrade the master branch code, and this feature will be released in the next version.

**Comment by Orionsg** [comment]
Many thanks, xuri. 
Works fine.

**Comment by liserjrqlxue** [comment]
I wonder why `MaxFileNameLength` is set to `207` in https://github.com/qax-os/excelize/blob/81d9362b4f1cf765712b61837d5b5831d1cd0c58/xmlDrawing.go#L106

It seems file name length limit is `255` in `golang`.

> To resolve this issue, we just set saved path in the [`SaveAs`](https://github.com/xuri/excelize/blob/544ef18a8cb9949fcb8833c6d2816783c90f3318/file.go#L67) like this:
> 
> ```go
> func (f *File) SaveAs(name string, opt ...Options) error {
>     f.Path = name
>     if len(name) > MaxFileNameLength {
>         return ErrMaxFileNameLength
>     }
>     ...
> }
> ```
> 
> Would you like to make PR for this? or I'll fix it later.

**Comment by xuri** [comment]
Accroding to [ISO 9660:1988](https://www.iso.org/standard/17505.html) International Standard §7.5.1, the Length of File Name shall not exceed 207.
